docs/proposals/20220725-managed-kubernetes.md

```diff
@@ -0,0 +1,493 @@
+
+- **GCPManagedControlPlane**: same as in option 1
+- **GCPManagedCluster**: contains the minimum properties in its spec and status to satisfy the [CAPI contract for an infrastructure cluster](https://cluster-api.sigs.k8s.io/developer/providers/cluster-infrastructure.html) (i.e. ControlPlaneEndpoint, Ready condition). Its controller watches GCPManagedControlPlane and copies the ControlPlaneEndpoint field to GCPManagedCluster to report back to CAPI. This is used as a pass-through layer only.
+
+```go
+type GCPManagedControlPlaneSpec struct {
+// Project is the name of the project to deploy the cluster to.
+Project string `json:"project"`
+
+// NetworkSpec encapsulates all things related to the GCP network.
+// +optional
+Network NetworkSpec `json:"network"`
+
+// AddonsConfig defines the addons to enable with the GKE cluster.
+// +optional
+AddonsConfig *AddonsConfig `json:"addonsConfig,omitempty"`
+
+// Logging contains the logging configuration for the GKE cluster.
+// +optional
+Logging *ControlPlaneLoggingSpec `json:"logging,omitempty"`
+
+// EnableKubernetesAlpha will indicate the kubernetes alpha features are enabled
+// +optional
+EnableKubernetesAlpha bool
+
+// ControlPlaneEndpoint represents the endpoint used to communicate with the control plane.
+// +optional
+ControlPlaneEndpoint clusterv1.APIEndpoint `json:"controlPlaneEndpoint"`
+....
+}
+```
+
+```go
+type GCPManagedClusterSpec struct {
+// ControlPlaneEndpoint represents the endpoint used to communicate with the control plane.
+// +optional
+ControlPlaneEndpoint clusterv1.APIEndpoint `json:"controlPlaneEndpoint"`
+}
+```
+
+**This is the design pattern used by AKS in CAPZ**. [An example of how ManagedCluster watches ControlPlane in CAPZ.](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/exp/controllers/azuremanagedcluster_controller.go#L165)
+
+**Pros**
+
+- Better aligned with CAPI’s traditional infra provider model
+- Works with ClusterClass
+
+**Cons**
+
+- Need to maintain Infra cluster kind, which is a pass-through layer and has no other functions. In addition to the CRD, controllers, webhooks and conversions webhooks need to be maintained.
+- Infra provider doesn’t provision infrastructure and whilst it may meet the CAPI contract, it doesn’t actually create infrastructure as this is done via the control plane.
+
+#### Option 3: Two kinds with a Managed Control Plane and Managed Infra Cluster with Better Separation of Responsibilities
+
+This option more closely follows the original separation of concerns with the different CAPI provider types. With this option, 2 new resource kinds will be introduced:
+
+- **GCPManagedControlPlane**: this presents the actual GKE control plane in GCP. Its spec would only contain properties that are specific to the provisioning & management of a GKE cluster in GCP (excluding worker nodes). It would not contain any properties related to the general GCP operating infrastructure, like the networking or project.
```

Comment:
I had discussed this in slack, but one observation with this approach is that the endpoint address is set the infra cluster object by the cluster reconciler, but in case of managed providers, the endpoint is created when the control plane is created.  So the cluster reconciler also needs to watch any changes to control plane and update the endpoint address when the control plane is completely initialised. Also if you look at https://cluster-api.sigs.k8s.io/developer/architecture/controllers/control-plane.html , their control planer reconciler checks for status.endpoint and only actions if that is set. In this case that wont work. Probably we need to give clear instructions to providers to achieve the same.

Reply:
Thanks for the comment, @shyamradhakrishnan. 

You are correct and `the endpoint is created when the control plane is created` is one of the major differences for managed k8s offerings. That observation is described in [Motivation](https://github.com/kubernetes-sigs/cluster-api/pull/6988/files#diff-bf3a5ea03d72b0d137e102bca815e6900a9e298627f527fde5161465a6e47f35R91-R92) section and it is why two existing implementations, CAPA (option1) and CAPZ (option2), ended up in two different approaches.

Summarizing discussion from [slack thread](https://kubernetes.slack.com/archives/C8TSNPY4T/p1657526148351049)
* The control plane reconcile needs to start when InfrastructureReady is true. How CAPZ is doing it ([code](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/exp/controllers/azuremanagedcontrolplane_controller.go#L104)).
* The managed cluster object will need to watch the control plane and report back when it has a control plane endpoint (whether the field is named the same or not is up to you).

We have [Provider Implementers Documentation](https://github.com/kubernetes-sigs/cluster-api/pull/6988/files#diff-bf3a5ea03d72b0d137e102bca815e6900a9e298627f527fde5161465a6e47f35R457) section to keep track of changes needed in the doc. Will add this item to it.

Reply:
Create a doc issue
* https://github.com/kubernetes-sigs/cluster-api/issues/7032